MathLib/LinAlg/SimpleMatrixVectorProvider-impl.h

```diff
@@ -0,0 +1,281 @@
+{
+    auto& ptr = it->first;
+    auto const id = it->second;
+
+    auto res = to_unused.emplace(id, ptr);
+    assert(res.second && "Emplacement failed.");
+    (void) res; // res unused if NDEBUG
+    from_used.erase(it);
+}
+
+} // detail
+
+
+namespace MathLib
+{
+
+template<typename Matrix, typename Vector>
+template<bool do_search, typename MatVec, typename... Args>
+std::pair<MatVec*, bool>
+SimpleMatrixVectorProvider<Matrix, Vector>::
+get_(std::size_t& id,
+     std::map<std::size_t, MatVec*>& unused_map,
+     std::map<MatVec*, std::size_t>& used_map,
+     Args&&... args)
+{
+    if (id >= _next_id) {
+        ERR("An obviously uninitialized id argument has been passed."
+            " This might not be a serious error for the current implementation,"
+            " but it might become one in the future."
+            " Hence, I will abort now.");
+        std::abort();
+    }
+
+    if (do_search)
+    {
+        auto it = unused_map.find(id);
+        if (it != unused_map.end()) // unused matrix/vector found
+            return { ::detail::transfer(unused_map, used_map, it), false };
+    }
+
+    // not searched or not found, so create a new one
+    id = _next_id++;
+    auto res = used_map.emplace(
+        MatrixVectorTraits<MatVec>::newInstance(std::forward<Args>(args)...).release(),
+        id);
+    assert(res.second && "Emplacement failed.");
+    return { res.first->first, true };
+}
+
+template<typename Matrix, typename Vector>
+template<bool do_search, typename... Args>
+std::pair<Matrix*, bool>
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix_(std::size_t& id, Args&&... args)
+{
+    return get_<do_search>(id, _unused_matrices, _used_matrices, std::forward<Args>(args)...);
+}
+
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix()
+{
+    std::size_t id = 0u;
+    return *getMatrix_<false>(id).first;
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(std::size_t& id)
+{
+    return *getMatrix_<true>(id).first;
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(MatrixSpecificationsProvider const& msp)
+{
+    std::size_t id = 0u;
+    auto const mat_spec = msp.getMatrixSpecifications();
+    return *getMatrix_<false>(id, mat_spec).first;
+    // TODO assert that the returned object always is of the right size
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(MatrixSpecificationsProvider const& msp, std::size_t& id)
+{
+    auto mat_spec = msp.getMatrixSpecifications();
+    return *getMatrix_<true>(id, mat_spec).first;
+    // TODO assert that the returned object always is of the right size
+}
+
+template<typename Matrix, typename Vector>
+Matrix&
+SimpleMatrixVectorProvider<Matrix, Vector>::
+getMatrix(Matrix const& A)
+{
+    std::size_t id = 0u;
```

Comment:
No, it cannot be reused. But there's a bug in the implementation of `getMatrix(id, A);`